content/en/blog/_posts/non-root-containers-and-devices.md

```diff
@@ -0,0 +1,238 @@
+both devices added and with non-root uid/gid set via `runAsUser`/`runAsGroup`: the container user process
+has no permission to use the device even when its group id (gid, copied from host) was permissive to
+non-root groups. This is because the container user does not belong to that host group (e.g., via `additionalGids`).
+
+Being able to run applications that use devices as non-root user is normal and expected to work so that
+the security principles can be met. Therefore, several alternatives were considered to get the gap filled with what the PodSec/CRI/OCI supports today.
+
+## What was done to solve the issue?
+You might have noticed from the problem definition that it would at least be possible to workaround
+the problem by manually adding the device gid(s) to `supplementalGroups`, or in
+the case of just one device, set `runAsGroup` to the device's group id. However, this is problematic because the device gid(s) may have
+different values depending on the nodes' distro/version in the cluster. For example, with GPUs the following commands for different distros and versions return different gids:
+
+Fedora 33:
+```
+$ ls -l /dev/dri/
+total 0
+drwxr-xr-x. 2 root root         80 19.10. 10:21 by-path
+crw-rw----+ 1 root video  226,   0 19.10. 10:42 card0
+crw-rw-rw-. 1 root render 226, 128 19.10. 10:21 renderD128
+$ grep -e video -e render /etc/group
+video:x:39:
+render:x:997:
+```
+
+Ubuntu 20.04:
+```
+$ ls -l /dev/dri/
+total 0
+drwxr-xr-x 2 root root         80 19.10. 17:36 by-path
+crw-rw---- 1 root video  226,   0 19.10. 17:36 card0
+crw-rw---- 1 root render 226, 128 19.10. 17:36 renderD128
+$ grep -e video -e render /etc/group
+video:x:44:
+render:x:133:
+```
+
+Which number to choose in your `securityContext`? Also, what if the `runAsGroup`/`runAsUser` values cannot be hard-coded because
+they are automatically assigned during pod admission time via external security policies?
+
+Unlike volumes with `fsGroup`, the devices have no official notion of `deviceGroup`/`deviceUser` that the CRI runtimes (or kubelet)
+would be able to use. We considered using container annotations set by the device plugins (e.g., `io.kubernetes.cri.hostDeviceSupplementalGroup/`) to get custom OCI `config.json` uid/gid values.
+This would have required changes to all existing device plugins which was not ideal.
+
+Instead, a solution that is *seamless* to end-users without getting the device plugin vendors involved was preferred. The selected approach was
+to re-use `runAsUser` and `runAsGroup` values in `config.json` for devices:
+
+```yaml
+{
+        "type": "c",
+        "path": "/dev/foo",
+        "major": 123,
+        "minor": 4,
+        "fileMode": 438,
+        "uid": <runAsUser>,
+        "gid": <runAsGroup>
+},
+```
+
+With `runc` OCI runtime (in non-[*rootless mode*](/docs/tasks/administer-cluster/kubelet-in-userns/)), the device is created (`mknod(2)`) in
```

Comment:
nit: I'm not sure I like the embedded link and text. 
With `runc` OCI runtime in non-rootless mode, [who or what creates the device?] the device is created (`mknod(2)`) in the container namespace and the ownership is changed to `runAsUser`/`runAsGroup` using `chmod(2)`.

Reply:
@kbhawkey thanks for the feedback, it makes sense. I wanted to keep the link to rootless so I moved it to the note below.